overlord/snapstate/handlers_components.go

```diff
@@ -227,3 +240,169 @@ func (m *SnapManager) undoMountComponent(t *state.Task, _ *tomb.Tomb) error {
 
 	return m.backend.RemoveComponentDir(cpi)
 }
+
+func (m *SnapManager) doLinkComponent(t *state.Task, _ *tomb.Tomb) error {
+	// invariant: component is not in the state (unlink happens previously if necessary)
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// snapSt is a copy of the current state
+	compSetup, snapsup, snapSt, err := compSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	csi := compSetup.CompSideInfo
+
+	// Grab information for current snap revision
+	// TODO will this still be correct when jointly installing snap +
+	// components? Link for the snap should happen earlier. But we do not
+	// want to reboot until all components are linked to the initramfs has
+	// the information to moun them already.
+	snapInfo, err := snapSt.CurrentInfo()
+	// the owner snap is expected to be installed (ErrNoCurrent not allowed)
+	if err != nil {
+		return err
+	}
+	snapIdx := snapSt.LastIndex(snapInfo.Revision)
+	revSt := snapSt.Sequence.Revisions[snapIdx]
+
+	// set information for undoLinkComponent in the task
+	t.Set("linked-component", csi)
+	// Append new component to components of the current snap
+	revSt.Components = append(revSt.Components, csi)
+
+	// Store last component refresh time
+	if snapSt.LastCompRefreshTime == nil {
+		snapSt.LastCompRefreshTime = make(map[string]time.Time)
+	}
+	snapSt.LastCompRefreshTime[csi.Component.ComponentName] = timeNow()
+
+	// Finally, write the state
+	Set(st, snapsup.InstanceName(), snapSt)
```

Comment:
we want to set Status to Done as we do in link-snap, otherwise this might be rerun and this code is not idempotent